### DISCUSSION TITLE
多线程安全问题

### DISCUSSION THREAD

**Discussion by Lj-dog** [BYJRK/DotNet-Discussions#313] [C# 技术交流]
看完视频了，但视频内容不好检索，且博客多线程内容没更新。所以在此提问。
  问题：多线程调用CPP的dll static方法崩溃。（Debug没有异常信息，直接闪退。从window日志看是这个CPP的dll模块错误，堆栈溢出）
  详细：主页页面使用Task.Run(Read)，Read里“调用 static方法读取数据并用 Dispatcher.Invoke更新UI内容”。
 想法：我是觉得有可能是没调用完这CPP的static方法就再次调用引起。我本来想用个async方法封装这CPP static方法但由于传染性和 听说**即使是async 不同线程同时调用也是可以进入。** 就作罢了。  
目前我只想到了 用lock （obj）{} 来锁住。  
不知道各位有没有碰到这情况，是怎么处理的呢

**Comment by Lj-dog** [comment]
现在发现使用lock后还是会出现堆栈溢出的情况。但是使用官方的demo（Demo里就直接开thread，Thread threadWrite = new System.Threading.Thread(new System.Threading.ThreadStart(Write));） 并没有出现闪退的情况？这如何确定问题点呢 要是wpf线程使用不合理至少VS会给出异常信息把

**Comment by BYJRK** [comment]
你或许可以考虑引入异步编程的同步机制，比如使用 `SemaphoreSlim` 来控制同时执行的异步任务的数量，或者使用 `BlockingCollection` 来将对于非托管方法的调用作为对象传入队列？
